core/src/main/java/com/linecorp/armeria/common/CompositeHttpHeadersBase.java

```diff
@@ -0,0 +1,212 @@
+        return builder.build();
+    }
+
+    private static List<StringMultimap<CharSequence, AsciiString>> from(List<HttpHeaderGetters> headers) {
+        if (headers.isEmpty()) {
+            return ImmutableList.of();
+        }
+
+        final ImmutableList.Builder<StringMultimap<CharSequence, AsciiString>> builder =
+                ImmutableList.builder();
+        for (HttpHeaderGetters header : headers) {
+            if (header instanceof HttpHeadersBase) {
+                builder.add((HttpHeadersBase) header);
+            } else {
+                builder.add(new HttpHeadersBase(header));
+            }
+        }
+
+        return builder.build();
+    }
+
+    final void contentLength(long contentLength) {
+        checkArgument(contentLength >= 0, "contentLength: %s (expected: >= 0)", contentLength);
+        remove0(HttpHeaderNames.CONTENT_LENGTH);
+        ((HttpHeadersBase) appender()).contentLength(contentLength);
+    }
+
+    @Override
+    public long contentLength() {
+        return getLong(HttpHeaderNames.CONTENT_LENGTH, -1);
+    }
+
+    final void contentLengthUnknown() {
+        remove0(HttpHeaderNames.CONTENT_LENGTH);
+        ((HttpHeadersBase) appender()).contentLengthUnknown();
+    }
+
+    @Override
+    public boolean isContentLengthUnknown() {
+        return ((HttpHeaderGetters) appender()).isContentLengthUnknown();
+    }
+
+    final void contentType(MediaType contentType) {
+        requireNonNull(contentType, "contentType");
+        remove0(HttpHeaderNames.CONTENT_TYPE);
+        ((HttpHeadersBase) appender()).contentType(contentType);
+    }
+
+    @Override
+    public @Nullable MediaType contentType() {
+        final String contentTypeString = get(HttpHeaderNames.CONTENT_TYPE);
+        if (contentTypeString == null) {
+            return null;
+        }
+
+        try {
+            return MediaType.parse(contentTypeString);
+        } catch (IllegalArgumentException ex) {
+            return null;
+        }
+    }
+
+    final void contentDisposition(ContentDisposition contentDisposition) {
+        requireNonNull(contentDisposition, "contentDisposition");
+        remove0(HttpHeaderNames.CONTENT_DISPOSITION);
+        ((HttpHeadersBase) appender()).contentDisposition(contentDisposition);
+    }
+
+    @Override
+    public @Nullable ContentDisposition contentDisposition() {
+        final String contentDispositionString = get(HttpHeaderNames.CONTENT_DISPOSITION);
+        if (contentDispositionString == null) {
+            return null;
+        }
+
+        try {
+            return ContentDisposition.parse(contentDispositionString);
+        } catch (IllegalArgumentException ex) {
+            return null;
+        }
+    }
+
+    final void endOfStream(boolean endOfStream) {
+        ((HttpHeadersBase) appender()).endOfStream(endOfStream);
+    }
+
+    @Override
+    public boolean isEndOfStream() {
+        for (StringMultimapGetters<CharSequence, AsciiString> delegate : delegates()) {
```

Comment:
Shouldn't we check the `appender` first?
Additionally, we need to ensure that the appender isn't created if it's null.

Reply:
> https://github.com/line/armeria/pull/5340#discussion_r1498488839

Reply:
> Additionally, we need to ensure that the appender isn't created if it's null.

```
    protected final StringMultimap<IN_NAME, NAME> appender() {
        if (appender != null) {
            return appender;
        }
        appender = requireNonNull(appenderSupplier.get(), "appender");

        final ImmutableList.Builder<StringMultimapGetters<IN_NAME, NAME>> builder = ImmutableList.builder();

        // 👈👈  If appender is created, add appender on delegates.
        delegates = builder.addAll(delegates()) 
                           .add(appender)
                           .build();
        return appender;
    }

    protected final List<StringMultimapGetters<IN_NAME, NAME>> delegates() {
        if (delegates != null) {  // 👈👈 if appender is created or delegates() is called before, return it.
            return delegates;
        }
    }
```
 Nice point! I considered this case and on `delegates()`, it doesn't contain `appender` if appender is not created yet~!

So we don't have to check appender is null or not when using `delegates()`!

Reply:
> As we discussed like above, I think it's not mandatory to always check appender first on isEndOfStream !

I apologize for my previous comment. It doesn't contain all the information
I thought we still needed to check the appender if `endOfStream` is set after the `CompositeHttpHeadersBase` is created.
I imagined a case something like:
```java
CompositeHttpHeadersBase headers =
  new CompositeHttpHeadersBase(additionalHeaders, endOfStreamTrueHeaders);
headers.endOfStream(false);
```